### PR TITLE
Synthesis Papercut: Error read-only and write-only memories

### LINKED ISSUE #384: Initial blocks for memories sometimes break synthesis
To solve the #288 (zero-time reset bug), we added `initial` blocks that sets every port to 0. However, for memories that are never used (like the input memories in the systolic array), writing `0` causes the memory not to be synthesized. For the systolic array, this can be solved by externalizing the memory interfaces for the input memories.

We should add a paper-cut lint that detects when non-externalized memories are never written to and throw an error when we try to synthesize.

### PR DESCRIPTION
**Pull request by rachitnigam**
New pass that generates and error when a memory is only used for reads or writes without @external(1). This pass should be enabled for the default synthesis alias.

Fixes #384.

### CODE REVIEW

File: calyx/src/passes/synthesis_papercut.rs [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+        let has_mem_parent =
+            |p: &ir::Port| memory_cells.contains(&p.get_parent_name());
+        let analysis =
+            GraphAnalysis::from(&*comp).edge_induced_subgraph(|p1, p2| {
+                has_mem_parent(p1) || has_mem_parent(p2)
+            });
+
+        for mem in memory_cells {
+            println!("{}", mem);
```
Comment (sgpthomas) [inline calyx/src/passes/synthesis_papercut.rs:72]: remove the debug print

File: calyx/src/passes/synthesis_papercut.rs [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+            let cell = comp.find_cell(&mem).unwrap();
+            let read_port = cell.borrow().get("read_data");
+            if analysis.reads_from(&*read_port.borrow()).next().is_none() {
+                return Err(Error::Papercut(
+                    format!(
+                        "Only reads performed on memory `{}'. Synthesis tools with remove this memory. Add @external(1) to cell to turn this into an interface memory.",
```
Comment (sgpthomas) [inline calyx/src/passes/synthesis_papercut.rs:78]: typo: 
> Synthesis tools ~with~*will* remove this memory

File: calyx/src/passes/synthesis_papercut.rs [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+                        mem.to_string()
+                    ),
+                    mem,
+                ));
+            }
+            let write_port = cell.borrow().get("write_data");
+            if analysis.writes_to(&*write_port.borrow()).next().is_none() {
+                return Err(Error::Papercut(
+                    format!(
+                        "Only writes performed on memory `{}'. Synthesis tools with remove this memory. Add @external(1) to cell to turn this into an interface memory.",
```
Comment (sgpthomas) [inline calyx/src/passes/synthesis_papercut.rs:88]: same typo: ~with~*will*

File: tests/errors/mem-only-reads.expect [hunk 4]
```diff
@@ -0,0 +1,7 @@
+mem
+---CODE---
+1
+---STDERR---
+Error: 
+4 |    mem = std_mem_d1(32, 4, 4);
+  |    ^^^ [Papercut] Only writes performed on memory `mem'. Synthesis tools with remove this memory. Add @external(1) to cell to turn this into an interface memory
```
Comment (sgpthomas) [inline tests/errors/mem-only-reads.expect:7]: this test was failing for some reason. any idea why?

**Review by sgpthomas — APPROVED** [review]
LGTM except for the typos and the failling test

File: tests/errors/mem-only-reads.expect (hunk 4 above)
Comment (rachitnigam) [reply tests/errors/mem-only-reads.expect]: Added `.` at the end. Forgot to update.

**Comment by rachitnigam** [comment]
Should be good now. Squash and merge when ready.

**Pull request merged**